plugins/org.python.pydev.ast/src/org/python/pydev/ast/codecompletion/revisited/visitors/FindDefinitionModelVisitor.java

```diff
@@ -391,6 +385,71 @@ public Object visitAssign(Assign node, int unpackPos) throws Exception {
         return super.visitAssign(node);
     }
 
+    /**
+     * @see org.python.pydev.parser.jython.ast.VisitorBase#visitIf(org.python.pydev.parser.jython.ast.If)
+     */
+    @Override
+    public Object visitIf(If node) throws Exception {
+        ILocalScope scope = new LocalScope(nature, this.defsStack);
+        scope.setFoundAtASTNode(node);
+        if (foundAsDefinition && !scope.equals(definitionFound.scope)) { //if it is found as a definition it is an 'exact' match, so, we do not keep checking it
+            return null;
+        }
+        getDefinitionForWalrus(node.body, node.test, scope);
+        return super.visitIf(node);
+    }
+
+    /**
+     * @see org.python.pydev.parser.jython.ast.VisitorBase#visitWhile(org.python.pydev.parser.jython.ast.While)
+     */
+    @Override
+    public Object visitWhile(While node) throws Exception {
+        ILocalScope scope = new LocalScope(nature, this.defsStack);
+        scope.setFoundAtASTNode(node);
+        if (foundAsDefinition && !scope.equals(definitionFound.scope)) { //if it is found as a definition it is an 'exact' match, so, we do not keep checking it
+            return null;
+        }
+        getDefinitionForWalrus(node.body, node.test, scope);
+        return super.visitWhile(node);
+    }
+
+    private boolean getDefinitionForWalrus(stmtType[] body, exprType test, ILocalScope scope) {
+        if (test == null || (!(test instanceof NamedExpr) && !(test instanceof Compare))) {
+            return false;
+        }
+
+        if (test instanceof Compare) {
+            Compare comp = (Compare) test;
+            if (comp.left == null) {
+                return false;
+            }
+            exprType[] elts = NodeUtils.extractElts(comp.left);
+            if (elts == null || elts.length != 1 || !(elts[0] instanceof NamedExpr)) {
+                return false;
+            }
+            test = elts[0];
+        }
+
+        for (stmtType content : body) {
```

Comment:
This is not ideal... You're effectively doing a manual `visitNamedExpr` here... Instead of doing that you should override the `visitNamedExpr` method and act upon it when it's reached... if needed you can do things in the `visitIf` or `visitWhile` to set things in the class so that the scope is correct when you arrive at the `visitNamedExpr` if you had need of information from them (although in this case I don't think you need it).